Strict types declaration fixer

Seldaek opened the pull request:
This adds a fixer that ensures all files start with

```
<?php declare(strict_types=1);

// ...
```

I also had to adjust the header_comment fixer because it shouldn't insert a header above the declare IMO.

The only problem I have is that the tests fail with:

```
4) PhpCsFixer\Tests\Fixer\Strict\StrictTypesFixerTest::testFixAddsToEmptyFile
The token at index 5 must be {
    "id": 317,
    "name": "T_LNUMBER",
    "content": 1,
    "isArray": true,
    "changed": false
}, got {
    "id": 317,
    "name": "T_LNUMBER",
    "content": 1,
    "isArray": true,
    "changed": true
}
Failed asserting that false is true.
```

I can't make sense of it though, as the code works and does the right thing but only this changed flag is true when it shouldn't or something. I don't know enough about internals to know how to fix this. Advice would be nice.

Another improvement I wanted to make is to allow a blacklist and/or whitelist to be configured on the fixer, like an array of files/regex to ignore or include, because this isn't strictly speaking coding style, it has runtime impact and you don't necessarily want to enforce this on ALL files. However isCandidate doesn't receive the file path, only the tokens, so it's hard to make a decision about the file at that point. I could do it in fix() I guess but wanted to ask first.

SpacePossum commented on src/Fixer/Strict/StrictTypesFixer.php:
```diff
@@ -0,0 +1,106 @@
+            new Token('('),
+            new Token(array(T_STRING, 'strict_types')),
+            new Token('='),
+            new Token(array(T_LNUMBER, 1)),
```
The `1` is the problem of the failing test. The content should be a string, i.e. change it to `'1'` and you should be OK :)

I'll add a note to my TODO list to make the output of the failing test better (i.e. it would be nice if the output of the `toJson` method of `Token` would show `'1'` as content and not `1`)

Seldaek replied:
OK thanks, fixed.

SpacePossum commented on tests/Fixer/Strict/StrictTypesFixerTest.php:
```diff
@@ -0,0 +1,105 @@
+    {
+        $input = " <?php\nphpinfo();\n";
+        $this->doTest($input);
+    }
```
Would be nice to have a test with a file with header, both in PHPDoc and comment block and using both. Just to see if and where the `declare` gets inserted.
Have a test for short tag open would be nice (not that short tag is nice, but it is not removed from PHP 7).

Seldaek replied:
Added a few tests to confirm all that.

GrahamCampbell commented:
Could we have an option to have it like this please:

![image](https://cloud.githubusercontent.com/assets/2829600/15609442/fbbca27c-2417-11e6-8ace-a7783e6a22d3.png)

GrahamCampbell commented:
More complete example of my prefered CS:

![image](https://cloud.githubusercontent.com/assets/2829600/15609486/3280b5be-2418-11e6-8838-1d2a1bc5ac5f.png)

Seldaek commented:
I knew this'd come up, but I'd say please no. Believe me you'll get over it soon enough and this avoids shuffling every single file 2 lines below when opened. Besides if you use this to ensure the whole project is in strict mode then hiding it up there makes sense IMO, as you know it's there.

GrahamCampbell commented:
Could you add test cases showing this moves this up to the correct place if it already exists please?

GrahamCampbell commented:
I'd still rather we could allow both standards though.

Seldaek commented:
It's already there https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1921/files#diff-effdfb27184a14e565c6d45e17fea171R22

GrahamCampbell commented:
:)

SpacePossum commented:
the detection and moving is not always correct, for example.
Input:
```php
<?php

            /*test*/declare(strict_types=1)/**/;echo 123;
```
gets fixed to:
```php
<?php declare(strict_types=1);

/*test*/declare(strict_types=1)/**/;echo 123;
```

(with regards of the PHP horror coders ^-^)

Seldaek commented:
Ah indeed, it only detects it at the beginning, it doesn't scan the whole file for it. That'll make things a lot slower I guess, but I can do it if you think it's needed.

SpacePossum commented:
It is sad, however we can break valid code (no matter how bad written it is) unless we mark the fixer `risky`.

The scan only needs to start at the index of `$tokens->getNextMeaningfulToken(0)`.
Since the first token must be an open tag (`isMonolithicPhp` ensures that). The next meaningful must be the `declare`, or otherwise a `declare` in the file is invalid (and doesn't pass linting), so you can stop the scan there.
If the `null === nextMeaningFulIndex` than it is an empty file, so there is a little short cut for that edge case.

SpacePossum commented:
and please add a test for this case as well:
```php
<? //test
  declare      (
strict_types  =1
)




?>
test
```

Seldaek commented:
OK I added a fix for the one appearing after some comments, and added a test for that last one you mentioned but IMO those tests are kinda useless as it's all non-monolithic PHP so it doesn't even get parsed.

SpacePossum commented:
@Seldaek thanks for all the work!
Please consider https://github.com/Seldaek/PHP-CS-Fixer/pull/1/files for the StrrictTypeFixer.
I've modified the fixer so it can handle more cases.

SpacePossum commented on src/Fixer/Strict/StrictTypesFixer.php:
```diff
@@ -0,0 +1,170 @@
+        end($sequenceLocation);
+        $sequenceEndIndex = $tokens->getNextMeaningfulToken(key($sequenceLocation));
+        if ($tokens[$sequenceEndIndex]->isGivenKind(T_CLOSE_TAG)) {
+            return; // don't move, comment placement might cause invalid code if moved
```
Just realized this can be moved as long as the close tag is left in place and a semicolon is added to the moved `declare` statement. I might be able to PR this later on this week.

gharlan commented on src/Fixer/Comment/HeaderCommentFixer.php:
```diff
@@ -126,43 +129,81 @@ private function findHeaderCommentIndex(Tokens $tokens)
+            }
+        }
+
+        return $isStrictTypes ? $index : $originalIndex;
```
I think this will pass your check (`$isStrictTypes` will be `true`):

```php
declare(ticks=1) {
    const strict_types = 1;
}
```
;)

SpacePossum replied:
is this valid? and if it s, what does it mean? ;)

gharlan replied:
hmm, the `const` is not valid there, but this is valid: https://3v4l.org/MZ2TH

keradus replied:
```php
declare(ticks=1);

const strict_types = 1;
```

gharlan replied:
@keradus but you example will not pass the check, because there is a `;` before `strict_types`.. ;)

keradus replied:
huh https://3v4l.org/Op6dm ?

gharlan replied:
I mean: `$isStrictTypes` will be `false` in your case.

gharlan commented on src/Fixer/Strict/StrictTypesFixer.php:
```diff
@@ -0,0 +1,170 @@
+        $sequence = $this->getDeclareStrictTypeSequence();
+        $searchIndex = $tokens->getNextMeaningfulToken(0);
+        if (null === $searchIndex) {
+            $this->insertNewSequence($tokens); // not found, insert declare statement
```
`$searchIndex` is `null`, so there is no meaningful token in file (which should not be happen because of `isCandidate` check). But then you add the declare sequence (without php open tag).
I think something is wrong here..

SpacePossum replied:
indeed `$searchIndex` can never be `null` here, the whole if case can be removed, good catch :+1:

gharlan replied:
hmm I was wrong!

```php
$searchIndex = $tokens->getNextMeaningfulToken(0);
```

I thought: if `$searchIndex` is `null`, there is no meaningful token in file.
But actually: there is no meaningful token after first token.

Seldaek commented:
@SpacePossum if you want just open a new PR based off this one.. so you can freely tinker with it :) That'll be easier for you and me I guess, because I don't have a ton of time to spend on this. 

I just wanted to convert monolog to strict mode which I have done using my simple patch, now if you wanna spend hours figuring out all the insane syntax people could use to break the fixer I'm sure that's best for the project but I don't have time or energy for this :) It could also be merged as is because it works, and then improved later as/if needed.

gharlan commented on src/Fixer/Strict/StrictTypesFixer.php:
```diff
@@ -0,0 +1,170 @@
+        if (false !== strpos($tokens[0]->getContent(), "\n")) {
+            $tokens[0]->setContent(trim($tokens[0]->getContent()).' ');
+        } elseif ($tokens[1]->isWhitespace()) {
+            $tokens[1]->setContent(' ');
```
I think we should remove the whitespace token to get `<?php declare..` (first space belongs to open tag token).

GrahamCampbell replied:
Also, what happens if someone puts a different type of declare there?

SpacePossum replied:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1921/files#diff-effdfb27184a14e565c6d45e17fea171R36

seems to work?

SpacePossum replied:
>Also, what happens if someone puts a different type of declare there?

Nothing happens to that declare, the strict type declare is moved to be the first statement after the open tag.

keradus commented:
This fixer should be marked as risky

SpacePossum commented:
>This fixer should be marked as risky

Wouldn't it be possible to make it non-risky?

keradus commented:
no, as `declare(strict_types=1)` changes how the code is executed

GrahamCampbell commented:
Yeh this is definitely risky. :)

SpacePossum commented:
I see, but that is the whole point :D I'm just not a big fan of warning people against themselves, but sure lets mark it risky :)

GrahamCampbell commented:
> I'm just not a big fan of warning people against themselves, but sure lets mark it risky :)

I think this is fine. People should know when a fixer is safe or not to avoid newcomers to php-cs-fixer from breaking their code by mistake.

keradus commented:
also, I pretty blindly run non-risky fixers against big codebase, and then activate risky fixers in smaller steps

Seldaek commented:
Closing in favor of https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1949
